loadtest/user/userentity/user.go

```diff
@@ -54,35 +58,56 @@ func New(store store.MutableUserStore, id int, config Config) *UserEntity {
 	}
 	ue.client.HttpClient = &http.Client{Transport: transport}
 	ue.store = store
+	ue.closing = make(chan struct{})
+	ue.closed = make(chan struct{})
 	return &ue
 }
 
+// Connect creates a websocket connection to the server and starts listening for messages.
 func (ue *UserEntity) Connect() error {
 	if ue.client.AuthToken == "" {
 		return errors.New("user is not authenticated")
 	}
-	if ue.wsClient != nil {
+	cli := ue.getWsClient()
+	if cli != nil {
 		return errors.New("user is already connected")
 	}
-	client, err := model.NewWebSocketClient4(ue.config.WebSocketURL, ue.client.AuthToken)
-	if err != nil {
-		return err
-	}
-	ue.wsClient = client
-	ue.wsClient.Listen()
-	// TODO: implement listener function
+
+	go ue.listen()
 	return nil
 }
 
+// Disconnect closes the websocket connection.
 func (ue *UserEntity) Disconnect() error {
-	if ue.wsClient == nil {
+	cli := ue.getWsClient()
+	if cli == nil {
+		return errors.New("user is not connected")
+	}
+	cli.Close()
+	cli = nil
+	// Wait for the listener to shut down.
+	close(ue.closing)
+	<-ue.closed
+	return nil
+}
+
+// SendWebsocketMessage sends a given action type with data to the websocket.
+func (ue *UserEntity) SendWebsocketMessage(action string, data map[string]interface{}) error {
+	cli := ue.getWsClient()
+	if cli == nil {
```

Comment:
Maybe we can do this comparison in the `getWsClient`? 2/5

Reply:
Actually, on the Connect method, we check for `cli != nil`, and now we are returning the error in a channel. So the code is different in different places.